app/pages/learn/01_tutorial/04_mastering-the-api/02_modern_io/index.md

```diff
@@ -0,0 +1,252 @@
+
+* Reading and writing text files
+* Reading text, images, JSON from the web
+* Visiting files in a directory
+* Reading a zip file
+* Creating a temporary file or directory
+
+The Java API supports many other tasks, which are explained in detail in the [Java I/O API tutorial](https://dev.java/learn/java-io/).
+
+Modern, at the time of this writing, means features that are out of preview in Java 21. In particular:
+
+* UTF-8 is the default for I/O since Java 18 ([JEP 400](https://openjdk.org/jeps/400))
+* The `java.nio.file.Files` class, which first appeared in Java 7, added useful methods in Java 8, 11, and 12
+* `java.io.InputStream` gained useful methods in Java 9, 11, and 12
+* The `java.io.File` and `java.io.BufferedReader` classes are now thoroughly obsolete, even though they appear frequently in web searches and AI chats.
+
+## Reading Text Files
+
+Tou can read a text file into a string like this:
+
+```
+String content = Files.readString(path);
+```
+
+Here, `path` is an instance of `java.nio.Path`, obtained like this:
+
+```
+var path = Path.of("/usr/share/dict/words");
+```
+
+If you want the file as a sequence of lines, call
+
+```
+List<String> lines = Files.readAllLines(path);
+```
+
+If the file is large, process the lines lazily as a `Stream<String>`:
+
+```
+try (Stream<String> lines = Files.lines(path)) 
+{
+   . . .
+}
+```
+
+Also use `Files.lines` if you can naturally process lines with stream operations (such as `map`, `filter`) .
+
+Note that the stream returned by `Files.lines` needs to be closed. To ensure that this happens, use a `try`-with-resources statement, as in the preceding code snippet.
+
+There is no longer a good reason to use the `readLine` method of `java.io.BufferedReader`.
+
+To split your input into something else than lines, use a `java.util.Scanner`. For example, here is how you can read words, separated by non-letters:
+
+```
+Stream<String> tokens = new Scanner(path).useDelimiter("\\PL+").tokens();
+```
+
+## Writing Text Files
+
+You can write a string to a text file with a single call:
+
+```
+String content = . . .;
+Files.writeString(path, content);
+```
+
+If you have a list of lines rather than a single string, use:
+
+```
+List<String> lines = . . .;
+Files.write(path, lines);
+```
+
+For more general output, use a `PrintWriter` so that you can use the `printf` method:
+
+```
+var writer = new PrintWriter(path.toFile());
+writer.printf("Hello, %s, next year you'll be %d years old!%n", name, age + 1);
+```
+
+Weirdly enough, as of Java 21, there is no `PrintWriter` constructor with a `Path` parameter.
+
+The `BufferedWriter` class can only write strings without formatting them. That is ok if you use the `String.formatted` method:
+
+```
+var writer = Files.newBufferedWriter(path);
+writer.write("Hello, %s, next year you'll be %d years old!%n".formatted(name, age + 1));
+```
+
+Or, with the `FMT` template (which is still in preview):
```

Comment:
It seems a bit inconsistent to mention preview features when `Modern, at the time of this writing, means features that are out of preview in Java 21.` was stated earlier.

Also, is the `FMT` processor really available by default in 21 (with previews enabled) or is it necessary to create a variable for that?

Reply:
Actually, since FMT is also locale-specific, I'll remove that.

Reply:
I wasn't really talking about the issue about being locale-specific but the fact that the following doesn't compile (with `--release 21 --enable-preview`):
```java
void main(){
	System.out.println(STR."Hi");//works
	System.out.println(FMT."Hi");//doesn't work
}
```
Instead, one needs to do it like this:
```java
import java.util.FormatProcessor;
import java.util.Locale;

final FormatProcessor FMT = FormatProcessor.create(Locale.ROOT);

void main(){
	System.out.println(STR."Hi");
	System.out.println(FMT."Hi");
}
```
As you can see, the later compiles and it is possible to specify a `Locale`.